src/main/java/me/oyurimatheus/nossomercadolivre/products/ProductOpinion.java

```diff
@@ -0,0 +1,83 @@
+package me.oyurimatheus.nossomercadolivre.products;
+
+import me.oyurimatheus.nossomercadolivre.users.User;
+import org.hibernate.validator.constraints.Length;
+import org.hibernate.validator.constraints.Range;
+
+import javax.persistence.*;
+import javax.validation.constraints.NotBlank;
+import javax.validation.constraints.NotNull;
+
+import static java.util.Objects.requireNonNull;
+import static javax.persistence.GenerationType.IDENTITY;
+import static org.springframework.util.Assert.hasText;
+
+@Table(name = "products_opinion")
+@Entity
+class ProductOpinion {
+
+    @Id
+    @GeneratedValue(strategy = IDENTITY)
+    private Long id;
+
+    @Range(min = 1, max = 5)
+    private Integer rating;
+
+    @NotBlank
+    private String title;
+
+    @Length(max = 500)
+    private String description;
+
+    @ManyToOne
+    @JoinColumn(name = "product_id")
+    @NotNull
+    private Product product;
+
+    @ManyToOne
+    @JoinColumn(name = "user_id")
+    @NotNull
+    private User user;
+
+    /**
+     * @deprecated  framework eyes only
+     */
+    @Deprecated
+    private ProductOpinion() { }
+
+    ProductOpinion(@Range(min = 1, max = 5) Integer rating,
```

Comment:
*Usamos tudo que conhecemos que está pronto. Só fazemos código do zero se for estritamente necessário.* + *Deixamos pistas que facilitem o uso do código onde não conseguimos resolver com compilação.*

Aqui foi legal o range e está suficiente na minha opinião. Por outro lado, você pode explorar a utilização de utilização de enum e garantir em tempo de compilação a restrição.

Reply:
Criar um `enum`, nesse caso, teria menos carga cognitiva do que um Integer? Ou é trade off justo para ter a garantia em compilação (não sei se ter essa garantia em compilação valeria a criação do `enum`)? 

Pensando que se pudesse mudar o formato do Range, agora é de 1 até 10 ou até 100 o quão complexo seria isso se utilizássemos um `enum`?

Reply:
Oi Yuri, levando em consideração a contagem exata do número de pontos de carga cognitiva, dado o meu ponto de vista, nem mudaria muito. Dado que você trocaria o if da validação da entrada do construtor pela criação de um arquivo novo. Mas podemos usar o cenário que você não precisaria criar o if dentro do construtor. Nesse cenário, a enum aumentaria sim a carga cognitiva. A contrapartida é que você eliminou um possível erro de runtime. 

Sobre ser de 1 a 10 ou de 1 a 100. A dificuldade é o trabalho braçal de adicionar a numeração. É como eu falei, apenas uma sugestão com uma outra visão :). O int cumpre seu papel, mas assim como no caso da senha encodada, o tipo que você quer aí não é simplesmente qualquer número, é um número carregado de sentido. 

Resumindo, com enum ou sem, ta tudo certo :). Queria só trazer mais uma perspectiva.

Reply:
Faz sentido! No final não só tipos primitivos. Eles têm semântica de fato. A ideia do enum para representar isso eu acho interessante 😄 . Só fico pensando se teria um jeito de não ter esse trabalho braçal de adicionar a numeração... mas aí só se mudasse a estrutura né?